File: _archive/archive/Lua/engine/libc/sdl/sdl.lua
local code, defs = Library.precompile(io.read('libc/sdl/sdl.c'))
ffi.cdef(code)

local loaded = pcall(loadstring('local _ = ffi.C.SDL_Init'))

class 'Sdl' : extends(Component) : meta(not loaded and Library.load('SDL2') or ffi.C)

function Sdl:init()
    self.window = NULL
    self.context = NULL
    return self
end

function Sdl:initialize()
    opengles = ios

    if opengles then
        config.glMajorVersion = 3
        config.glMinorVersion = 2
    else
        config.glMajorVersion = 4
        config.glMinorVersion = 1
    end

    if love then
        self.window = sdl.SDL_GL_GetCurrentWindow()
        if self.window ~= NULL then
            self.context = sdl.SDL_GL_GetCurrentContext()
            if self.context ~= NULL then
            end
        end
    end

    if self.window == NULL then
        if self.SDL_Init(self.SDL_INIT_EVERYTHING) == 0 then
            self:setCursor(sdl.SDL_SYSTEM_CURSOR_WAIT)

            if self.SDL_GL_LoadLibrary(ffi.NULL) == 1 then
                self.SDL_Log("SDL_GL_LoadLibrary: %s", self.SDL_GetError())
                error('SDL_GL_LoadLibrary')
            end

            if opengles then
                sdl.SDL_GL_SetAttribute(sdl.SDL_GL_CONTEXT_PROFILE_MASK, sdl.SDL_GL_CONTEXT_PROFILE_ES)

            else
                if config.glMajorVersion == 4 then
                    self.SDL_GL_SetAttribute(self.SDL_GL_CONTEXT_PROFILE_MASK, self.SDL_GL_CONTEXT_PROFILE_CORE)

                else
                    config.glMajorVersion = 3
                    config.glMinorVersion = 1

                    self.SDL_GL_SetAttribute(self.SDL_GL_CONTEXT_PROFILE_MASK, self.SDL_GL_CONTEXT_PROFILE_COMPATIBILITY)
                end
            end

            self.SDL_GL_SetAttribute(self.SDL_GL_CONTEXT_MAJOR_VERSION, config.glMajorVersion)
            self.SDL_GL_SetAttribute(self.SDL_GL_CONTEXT_MINOR_VERSION, config.glMinorVersion)

            self.SDL_GL_SetAttribute(self.SDL_GL_DOUBLEBUFFER, 1)
            self.SDL_GL_SetAttribute(self.SDL_GL_DEPTH_SIZE, 24)

            window = self.SDL_CreateWindow('Engine',
                0, 0,
                0, 0,
                self.SDL_WINDOW_OPENGL +
                --                self.SDL_WINDOW_ALLOW_HIGHDPI +
                --                self.SDL_WINDOW_FULLSCREEN +
                --                self.SDL_WINDOW_SHOWN +
                --                self.SDL_WINDOW_BORDERLESS +
                --                self.SDL_WINDOW_RESIZABLE +
                --                self.SDL_WINDOW_MAXIMIZED +
                0)

            if window then
                context = self.SDL_GL_CreateContext(window)
                if context then
                    self.window = window
                    self.context = context
                end
            end
        end
    end

    if self.window ~= NULL then
        local size = ffi.new('SDL_Rect')
        local displayIndex = 0

        if self.SDL_GetDisplayBounds(displayIndex, size) ~= 0 then
            self.SDL_Log("SDL_GetDisplayBounds failed: %s", self.SDL_GetError())
        end

        print(ffi.string(sdl.SDL_GetDisplayName(displayIndex)))

        -- TODEL
--        self:setWindowSize(screen)

        if self.context ~= NULL then
            local res = self.SDL_GL_MakeCurrent(self.window, self.context)
            assert(res == 0)

            -- 0 for immediate updates, 1 for updates synchronized with the vertical retrace, -1 for adaptive vsync
            self.SDL_GL_SetSwapInterval(1)

            local ddpi, hdpi, vdpi = ffi.new('float[1]'), ffi.new('float[1]'), ffi.new('float[1]')
            self.SDL_GetDisplayDPI(displayIndex,
                ddpi,
                hdpi,
                vdpi)

            self.ddpi, self.hdpi, self.vdpi = ddpi[0], hdpi[0], vdpi[0]
        end
    end

    sdl.image = class 'SdlImage' : meta(windows and Library.load('SDL2_image') or ffi.C)
end

function Sdl:setWindowSize(screen)
    self.SDL_HideWindow(self.window)
    do
--        self.SDL_MaximizeWindow(self.window)

        self.SDL_SetWindowSize(self.window,
            screen.w,
            screen.h)

        self.SDL_SetWindowPosition(self.window,
            100,
            100) --sdl.SDL_WINDOWPOS_CENTERED, sdl.SDL_WINDOWPOS_CENTERED)
    end

    self.SDL_ShowWindow(self.window)
end

function Sdl:setCursor(cursorId)
    local cursor = sdl.SDL_CreateSystemCursor(cursorId)
    sdl.SDL_SetCursor(cursor)
end

function Sdl:toggleWindowDisplayMode()
    sdl.SDL_SetWindowFullscreen(sdl.window, sdl.SDL_WINDOW_FULLSCREEN)
end

function Sdl:release()
    if self.context then
        self.SDL_GL_DeleteContext(self.context)
        self.context = NULL

        if self.window then
            self.SDL_DestroyWindow(self.window)
            self.window = NULL
        end
        self.SDL_GL_UnloadLibrary()
    end
    self.SDL_Quit()
end

function Sdl:scancode2key(event)
    local key = ffi.string(self.SDL_GetScancodeName(event.key.keysym.scancode))
    if key:len() <= 1 then
        key = ffi.string(self.SDL_GetKeyName(event.key.keysym.sym))
    end
    key = key:lower()
    key = mapKey(key, true)
    return key
end

local event = ffi.new('SDL_Event')
function Sdl:event()
    while self.SDL_PollEvent(event) == 1 do
        if event.type == self.SDL_WINDOWEVENT then
            if event.window.event == self.SDL_WINDOWEVENT_CLOSE then
                engine:quit()

            elseif event.window.event == sdl.SDL_WINDOWEVENT_RESIZED then
                -- TODO
            end

        elseif event.type == self.SDL_KEYDOWN or event.type == sdl.SDL_TEXTINPUT then
            engine:keydown(self:scancode2key(event), event.key.isrepeat)

        elseif event.type == sdl.SDL_KEYUP then
            engine:keyup(self:scancode2key(event), event.key.isrepeat)

        elseif event.type == sdl.SDL_MOUSEBUTTONDOWN then
            if event.button.button == self.SDL_BUTTON_LEFT then
                mouse:mouseEvent(
                    0,
                    BEGAN,
                    event.button.x, event.button.y,
                    0, 0,
                    true,
                    event.button.clicks)
            else
                engine:buttondown(event.button.button)
            end

        elseif event.type == sdl.SDL_MOUSEMOTION then
            local isTouch = bitAND(event.motion.state, self.SDL_BUTTON_LMASK)
            if isTouch then
                mouse:mouseEvent(
                    0,
                    MOVING,
                    event.motion.x, event.motion.y,
                    event.motion.xrel, event.motion.yrel,
                    isTouch)
            else
                mouse:mouseMove(
                    0,
                    MOVING,
                    event.motion.x, event.motion.y,
                    event.motion.xrel, event.motion.yrel,
                    isTouch)
            end

        elseif event.type == sdl.SDL_MOUSEBUTTONUP then
            if event.button.button == self.SDL_BUTTON_LEFT then
                mouse:mouseEvent(
                    0,
                    ENDED,
                    event.button.x, event.button.y,
                    0, 0,
                    false)
            else
                engine:buttonup(event.button.button)
            end

        elseif event.type == sdl.SDL_MOUSEWHEEL  then
            mouse:mouseWheel(1, event.wheel.x, event.wheel.y)

        elseif event.type == sdl.SDL_MULTIGESTURE then
            if event.mgesture.numFingers == 3 then
                engine:managerApp()
            end

        elseif event.type == sdl.SDL_FINGERDOWN then
        end
    end
end

function Sdl:swap()
    self.SDL_GL_SwapWindow(self.window)
end

File: _archive/archive/Lua/engine/libc/sdl/sdl.c
typedef unsigned char Uint8;
typedef unsigned short Uint16;
typedef unsigned int Uint32;

typedef signed char Sint8;
typedef signed short Sint16;
typedef signed int Sint32;
typedef int64_t Sint64;

typedef enum {
    SDL_INIT_TIMER = 0x00000001,
    SDL_INIT_AUDIO = 0x00000010,
    SDL_INIT_VIDEO = 0x00000020,
    SDL_INIT_JOYSTICK = 0x00000200,
    SDL_INIT_HAPTIC = 0x00001000,
    SDL_INIT_GAMECONTROLLER = 0x00002000,
    SDL_INIT_EVENTS = 0x00004000,
    SDL_INIT_NOPARACHUTE = 0x00100000,
    SDL_INIT_EVERYTHING = ( SDL_INIT_VIDEO | SDL_INIT_EVENTS )
} SDL_Flag;

typedef enum {
    SDL_WINDOW_FULLSCREEN = 0x00000001,
    SDL_WINDOW_OPENGL = 0x00000002,
    SDL_WINDOW_SHOWN = 0x00000004,
    SDL_WINDOW_HIDDEN = 0x00000008,
    SDL_WINDOW_BORDERLESS = 0x00000010,
    SDL_WINDOW_RESIZABLE = 0x00000020,
    SDL_WINDOW_MINIMIZED = 0x00000040,
    SDL_WINDOW_MAXIMIZED = 0x00000080,
    SDL_WINDOW_INPUT_GRABBED = 0x00000100,
    SDL_WINDOW_INPUT_FOCUS = 0x00000200,
    SDL_WINDOW_MOUSE_FOCUS = 0x00000400,
    SDL_WINDOW_FULLSCREEN_DESKTOP = ( SDL_WINDOW_FULLSCREEN | 0x00001000 ),
    SDL_WINDOW_FOREIGN = 0x00000800,
    SDL_WINDOW_ALLOW_HIGHDPI = 0x00002000
} SDL_WindowFlags;

#define SDL_WINDOWPOS_CENTERED 0x2FFF0000u
//#define SDL_WINDOWPOS_CENTERED SDL_WINDOWPOS_CENTERED_DISPLAY(0)
//#define SDL_WINDOWPOS_CENTERED_DISPLAY(X) (SDL_WINDOWPOS_CENTERED_MASK|(X))
//#define SDL_WINDOWPOS_CENTERED_MASK 0x2FFF0000u

int SDL_Init(Uint32 flags);
void SDL_Quit(void);

void SDL_Delay(Uint32 ms);

typedef enum {
    SDL_THREAD_PRIORITY_LOW,
    SDL_THREAD_PRIORITY_NORMAL,
    SDL_THREAD_PRIORITY_HIGH
} SDL_ThreadPriority;

int SDL_SetThreadPriority(SDL_ThreadPriority priority);

int SDL_GL_SetSwapInterval(int interval);
int SDL_GL_GetSwapInterval(void);

Uint32 SDL_GetTicks(void);

char* SDL_GetError();
void SDL_Log(const char *fmt, ...);

typedef void* SDL_Window;
typedef void* SDL_GLContext;

typedef enum {
    SDL_GL_RED_SIZE,
    SDL_GL_GREEN_SIZE,
    SDL_GL_BLUE_SIZE,
    SDL_GL_ALPHA_SIZE,
    SDL_GL_BUFFER_SIZE,
    SDL_GL_DOUBLEBUFFER,
    SDL_GL_DEPTH_SIZE,
    SDL_GL_STENCIL_SIZE,
    SDL_GL_ACCUM_RED_SIZE,
    SDL_GL_ACCUM_GREEN_SIZE,
    SDL_GL_ACCUM_BLUE_SIZE,
    SDL_GL_ACCUM_ALPHA_SIZE,
    SDL_GL_STEREO,
    SDL_GL_MULTISAMPLEBUFFERS,
    SDL_GL_MULTISAMPLESAMPLES,
    SDL_GL_ACCELERATED_VISUAL,
    SDL_GL_RETAINED_BACKING,
    SDL_GL_CONTEXT_MAJOR_VERSION,
    SDL_GL_CONTEXT_MINOR_VERSION,
    SDL_GL_CONTEXT_EGL,
    SDL_GL_CONTEXT_FLAGS,
    SDL_GL_CONTEXT_PROFILE_MASK,
    SDL_GL_SHARE_WITH_CURRENT_CONTEXT,
    SDL_GL_FRAMEBUFFER_SRGB_CAPABLE
} SDL_GLattr;

typedef enum
{
    SDL_GL_CONTEXT_PROFILE_CORE = 0x0001,
    SDL_GL_CONTEXT_PROFILE_COMPATIBILITY = 0x0002,
    SDL_GL_CONTEXT_PROFILE_ES = 0x0004
} SDL_GLprofile;

SDL_Window* SDL_CreateWindow(const char* title, int x, int y, int w, int h, Uint32 flags);
SDL_Window* SDL_GL_GetCurrentWindow(void);
void SDL_DestroyWindow(SDL_Window* window);

Uint32 SDL_GetWindowFlags(SDL_Window* window);

int SDL_SetWindowFullscreen(SDL_Window* window, Uint32 flags);
int SDL_SetWindowOpacity(SDL_Window* window, float opacity);

void SDL_SetWindowPosition(SDL_Window* window, int x, int y);
void SDL_SetWindowSize(SDL_Window* window, int w, int h);
void SDL_SetWindowTitle(SDL_Window* window, const char* title);

const char* SDL_GetWindowTitle(SDL_Window* window);

void SDL_ShowWindow(SDL_Window* window);
void SDL_RaiseWindow(SDL_Window* window);
void SDL_HideWindow(SDL_Window* window);

void SDL_MaximizeWindow(SDL_Window* window);
void SDL_MinimizeWindow(SDL_Window* window);
void SDL_RestoreWindow(SDL_Window* window);

int SDL_GL_LoadLibrary(const char* path);
void* SDL_GL_GetProcAddress(const char* proc);
void SDL_GL_UnloadLibrary(void);

void SDL_GL_ResetAttributes(void);
int SDL_GL_SetAttribute(SDL_GLattr attr, int value);
int SDL_GL_GetAttribute(SDL_GLattr attr, int *value);

SDL_GLContext SDL_GL_CreateContext(SDL_Window* window);
SDL_GLContext SDL_GL_GetCurrentContext(void);
int SDL_GL_MakeCurrent(SDL_Window* window, SDL_GLContext context);
void SDL_GL_SwapWindow(SDL_Window* window);
void SDL_GL_DeleteContext(SDL_GLContext context);

typedef void* SDL_Renderer;
SDL_Renderer* SDL_CreateRenderer(SDL_Window* window, int index, Uint32 flags);
void SDL_DestroyRenderer(SDL_Renderer* renderer);

typedef enum {
    SDL_FIRSTEVENT = 0,
    SDL_QUIT = 0x100,
    SDL_APP_TERMINATING,
    SDL_APP_LOWMEMORY,
    SDL_APP_WILLENTERBACKGROUND,
    SDL_APP_DIDENTERBACKGROUND,
    SDL_APP_WILLENTERFOREGROUND,
    SDL_APP_DIDENTERFOREGROUND,
    SDL_WINDOWEVENT = 0x200,
    SDL_SYSWMEVENT,
    SDL_KEYDOWN = 0x300,
    SDL_KEYUP,
    SDL_TEXTEDITING,
    SDL_TEXTINPUT,
    SDL_MOUSEMOTION = 0x400,
    SDL_MOUSEBUTTONDOWN,
    SDL_MOUSEBUTTONUP,
    SDL_MOUSEWHEEL,
    SDL_JOYAXISMOTION = 0x600,
    SDL_JOYBALLMOTION,
    SDL_JOYHATMOTION,
    SDL_JOYBUTTONDOWN,
    SDL_JOYBUTTONUP,
    SDL_JOYDEVICEADDED,
    SDL_JOYDEVICEREMOVED,
    SDL_CONTROLLERAXISMOTION = 0x650,
    SDL_CONTROLLERBUTTONDOWN,
    SDL_CONTROLLERBUTTONUP,
    SDL_CONTROLLERDEVICEADDED,
    SDL_CONTROLLERDEVICEREMOVED,
    SDL_CONTROLLERDEVICEREMAPPED,
    SDL_FINGERDOWN = 0x700,
    SDL_FINGERUP,
    SDL_FINGERMOTION,
    SDL_DOLLARGESTURE = 0x800,
    SDL_DOLLARRECORD,
    SDL_MULTIGESTURE,
    SDL_CLIPBOARDUPDATE = 0x900,
    SDL_DROPFILE = 0x1000,
    SDL_RENDER_TARGETS_RESET = 0x2000,
    SDL_USEREVENT = 0x8000,
    SDL_LASTEVENT = 0xFFFF
} SDL_EventType;

typedef enum {
    SDL_WINDOWEVENT_NONE,
    SDL_WINDOWEVENT_SHOWN,
    SDL_WINDOWEVENT_HIDDEN,
    SDL_WINDOWEVENT_EXPOSED,

    SDL_WINDOWEVENT_MOVED,

    SDL_WINDOWEVENT_RESIZED,
    SDL_WINDOWEVENT_SIZE_CHANGED,
    SDL_WINDOWEVENT_MINIMIZED,
    SDL_WINDOWEVENT_MAXIMIZED,
    SDL_WINDOWEVENT_RESTORED,

    SDL_WINDOWEVENT_ENTER,
    SDL_WINDOWEVENT_LEAVE,
    SDL_WINDOWEVENT_FOCUS_GAINED,
    SDL_WINDOWEVENT_FOCUS_LOST,
    SDL_WINDOWEVENT_CLOSE
} SDL_WindowEventID;

typedef struct SDL_WindowEvent {
    Uint32 type;
    Uint32 timestamp;
    Uint32 windowID;
    Uint8 event;
    Uint8 padding1;
    Uint8 padding2;
    Uint8 padding3;
    Sint32 data1;
    Sint32 data2;
} SDL_WindowEvent;

typedef struct SDL_MouseMotionEvent {
    Uint32 type;
    Uint32 timestamp;
    Uint32 windowID;
    Uint32 which;
    Uint32 state;
    Sint32 x;
    Sint32 y;
    Sint32 xrel;
    Sint32 yrel;
} SDL_MouseMotionEvent;

#define SDL_BUTTON_LMASK 1
#define SDL_BUTTON_MMASK 2
#define SDL_BUTTON_RMASK 4

#define SDL_BUTTON_LEFT 1
#define SDL_BUTTON_MIDDLE 2
#define SDL_BUTTON_RIGHT 3
#define SDL_BUTTON_X1 4
#define SDL_BUTTON_X2 5

typedef Sint32 SDL_Scancode;
typedef Sint32 SDL_Keycode;

typedef enum {
    KMOD_NONE = 0x0000,
    KMOD_LSHIFT = 0x0001,
    KMOD_RSHIFT = 0x0002,
    KMOD_LCTRL = 0x0040,
    KMOD_RCTRL = 0x0080,
    KMOD_LALT = 0x0100,
    KMOD_RALT = 0x0200,
    KMOD_LGUI = 0x0400,
    KMOD_RGUI = 0x0800,
    KMOD_NUM = 0x1000,
    KMOD_CAPS = 0x2000,
    KMOD_MODE = 0x4000,
    KMOD_RESERVED = 0x8000
} SDL_Keymod;

typedef struct SDL_Keysym {
    SDL_Scancode scancode;
    SDL_Keycode sym;
    Uint16 mod;
    Uint32 unused;
} SDL_Keysym;

typedef struct SDL_KeyboardEvent {
    Uint32 type;
    Uint32 timestamp;
    Uint32 windowID;
    Uint8 state;
    Uint8 isrepeat;
    Uint8 padding2;
    Uint8 padding3;
    SDL_Keysym keysym;
} SDL_KeyboardEvent;

typedef Sint64 SDL_TouchID;
typedef Sint64 SDL_FingerID;

typedef struct SDL_TouchFingerEvent {
    Uint32 type;
    Uint32 timestamp;
    SDL_TouchID touchId;
    SDL_FingerID fingerId;
    float x;
    float y;
    float dx;
    float dy;
    float pressure;
    Uint32 windowID;
} SDL_TouchFingerEvent;

typedef struct SDL_MouseButtonEvent {
    Uint32 type;
    Uint32 timestamp;
    Uint32 windowID;
    Uint32 which;
    Uint8 button;
    Uint8 state;
    Uint8 clicks;
    Uint8 padding1;
    Sint32 x;
    Sint32 y;
} SDL_MouseButtonEvent;

typedef struct SDL_MouseWheelEvent {
    Uint32 type;
    Uint32 timestamp;
    Uint32 windowID;
    Uint32 which;
    Sint32 x;
    Sint32 y;
    Uint32 direction;
} SDL_MouseWheelEvent;

typedef struct SDL_MultiGestureEvent {
    Uint32 type;
    Uint32 timestamp;
    SDL_TouchID touchId;
    float dTheta;
    float dDist;
    float x;
    float y;
    Uint16 numFingers;
    Uint16 padding;
} SDL_MultiGestureEvent;

typedef union SDL_Event {
    Uint32 type;
    SDL_WindowEvent window;
    SDL_MouseMotionEvent motion;
    SDL_KeyboardEvent key;
    SDL_TouchFingerEvent tfinger;
    SDL_MouseButtonEvent button;
    SDL_MouseWheelEvent wheel;
    SDL_MultiGestureEvent mgesture;
    Uint8 padding[56];
} SDL_Event;

void SDL_SetModState(SDL_Keymod modstate);
SDL_Keycode SDL_GetKeyFromScancode(SDL_Scancode scancode);
SDL_Scancode SDL_GetScancodeFromKey(SDL_Keycode key);
const char * SDL_GetScancodeName(SDL_Scancode scancode);
SDL_Scancode SDL_GetScancodeFromName(const char *name);
const char * SDL_GetKeyName(SDL_Keycode key);
SDL_Keycode SDL_GetKeyFromName(const char *name);
void SDL_StartTextInput(void);
const Uint8* SDL_GetKeyboardState(int* numkeys);

typedef struct SDL_Point {
    int x;
    int y;
} SDL_Point;

typedef struct SDL_Rect {
    int x, y;
    int w, h;
} SDL_Rect;

int SDL_GetNumVideoDisplays(void);

int SDL_GetDisplayBounds(int displayIndex, SDL_Rect* rect);
int SDL_GetDisplayDPI(int displayIndex, float* ddpi, float* hdpi, float* vdpi);
const char* SDL_GetDisplayName(int displayIndex);

int SDL_PollEvent(SDL_Event* event);

int SDL_SetRenderDrawColor(SDL_Renderer* renderer, Uint8 r, Uint8 g, Uint8 b, Uint8 a);
int SDL_RenderClear(SDL_Renderer* renderer);

int SDL_RenderDrawPoint(SDL_Renderer* renderer, int x, int y);
int SDL_RenderDrawPoints(SDL_Renderer* renderer, const SDL_Point* points, int count);
int SDL_RenderDrawLine(SDL_Renderer* renderer, int x1, int y1, int x2, int y2);
int SDL_RenderDrawLines(SDL_Renderer* renderer, const SDL_Point* points, int count);
int SDL_RenderDrawRect(SDL_Renderer* renderer, const SDL_Rect* rect);
int SDL_RenderFillRect(SDL_Renderer* renderer, const SDL_Rect* rect);

void SDL_RenderPresent(SDL_Renderer* renderer);

Uint32 SDL_GetMouseState(int *x, int *y);

typedef struct SDL_Color {
    Uint8 r;
    Uint8 g;
    Uint8 b;
    Uint8 a;
} SDL_Color;

typedef struct SDL_Palette {
    int ncolors;
    SDL_Color *colors;
    Uint32 version;
    int refcount;
} SDL_Palette;

typedef struct SDL_PixelFormat {
    Uint32 format;
    SDL_Palette *palette;
    Uint8 BitsPerPixel;
    Uint8 BytesPerPixel;
    Uint8 padding[2];
    Uint32 Rmask;
    Uint32 Gmask;
    Uint32 Bmask;
    Uint32 Amask;
    Uint8 Rloss;
    Uint8 Gloss;
    Uint8 Bloss;
    Uint8 Aloss;
    Uint8 Rshift;
    Uint8 Gshift;
    Uint8 Bshift;
    Uint8 Ashift;
    int refcount;
    struct SDL_PixelFormat *next;
} SDL_PixelFormat;

typedef struct SDL_Surface {
    Uint32 flags;
    SDL_PixelFormat *format;
    int w, h;
    int pitch;
    void *pixels;

    void *userdata;

    int locked;
    void *lock_data;

    SDL_Rect clip_rect;

    struct SDL_BlitMap *map;

    int refcount;
} SDL_Surface;

SDL_Surface* IMG_Load(const char *file);
SDL_Surface* SDL_CreateRGBSurfaceFrom(void *pixels,
    int width,
    int height,
    int depth,
    int pitch,
    Uint32 Rmask,
    Uint32 Gmask,
    Uint32 Bmask,
    Uint32 Amask);

int SDL_LockSurface(SDL_Surface *surface);
void SDL_UnlockSurface(SDL_Surface *surface);

void SDL_FreeSurface(SDL_Surface *surface);

int IMG_SavePNG(SDL_Surface *surface, const char *file);
int IMG_SaveJPG(SDL_Surface *surface, const char *file, int quality);

typedef struct SDL_Cursor SDL_Cursor;   /**< Implementation dependent */

typedef enum {
    SDL_SYSTEM_CURSOR_ARROW,     /**< Arrow */
    SDL_SYSTEM_CURSOR_IBEAM,     /**< I-beam */
    SDL_SYSTEM_CURSOR_WAIT,      /**< Wait */
    SDL_SYSTEM_CURSOR_CROSSHAIR, /**< Crosshair */
    SDL_SYSTEM_CURSOR_WAITARROW, /**< Small wait cursor (or Wait if not available) */
    SDL_SYSTEM_CURSOR_SIZENWSE,  /**< Double arrow pointing northwest and southeast */
    SDL_SYSTEM_CURSOR_SIZENESW,  /**< Double arrow pointing northeast and southwest */
    SDL_SYSTEM_CURSOR_SIZEWE,    /**< Double arrow pointing west and east */
    SDL_SYSTEM_CURSOR_SIZENS,    /**< Double arrow pointing north and south */
    SDL_SYSTEM_CURSOR_SIZEALL,   /**< Four pointed arrow pointing north, south, east, and west */
    SDL_SYSTEM_CURSOR_NO,        /**< Slashed circle or crossbones */
    SDL_SYSTEM_CURSOR_HAND,      /**< Hand */
    SDL_NUM_SYSTEM_CURSORS
} SDL_SystemCursor;

SDL_Cursor* SDL_CreateSystemCursor(SDL_SystemCursor id);
void SDL_SetCursor(SDL_Cursor* cursor);
